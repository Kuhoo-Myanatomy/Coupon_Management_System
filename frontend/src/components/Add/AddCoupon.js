import React, { Component } from 'react';
import './add.css'
import PropTypes from 'prop-types'

class AddCoupon extends Component{
    constructor(props){
        super(props)
        this.state={
            _id:this.props._id,
            couponCode:this.props.couponCode,
            discount:this.props.discount,
            issueDate:this.props.issueDate,
            expiryDate:this.props.expiryDate,
            eupdate:false,
        }
    }
    onChange = (e) => {
        console.log("Onchange");
        this.setState({ 
            [e.target.name]: e.target.value });
      };

    editCoupon=(event)=>{
        // event.preventDefault();
        console.log("Edit Method called");
        console.log(this.state);
        const {_id}=this.props;
        console.log(_id);
        const{ couponCode,discount,issueDate,expiryDate}=this.state;
        console.log(_id,couponCode,discount,issueDate,expiryDate);
        console.log(this.state);
        this.props.updateCoupon({_id,couponCode,discount,issueDate,expiryDate});
        // this.props.history.push('/coupons');
    }
    saveCoupon=(event)=>{
        event.preventDefault();
        console.log("Save Method called");

        console.log(this.state);
        const{ couponCode,discount,issueDate,expiryDate}=this.state;
        console.log(couponCode,discount,issueDate,expiryDate);
        this.props.createCoupon({couponCode,discount,issueDate,expiryDate});
    }
    render(){
        let heading="Edit ";
        if(!this.props.eupdate){
            // eupdate=false;
            heading="Add ";
        }
        var{ _id,couponCode,discount,issueDate,expiryDate,eupdate}=this.state
        const idate=new Date(issueDate),
        mnth = ("0" + (idate.getMonth() + 1)).slice(-2),
        day = ("0" + idate.getDate()).slice(-2);
        issueDate= [idate.getFullYear(), mnth, day].join("-");

        const edate=new Date(expiryDate),
        emnth = ("0" + (edate.getMonth() + 1)).slice(-2),
        eday = ("0" + edate.getDate()).slice(-2);
        expiryDate= [edate.getFullYear(), emnth, eday].join("-");
        
        console.log(typeof expiryDate,expiryDate);
        
        return (
            
            <div className='box'>
                <form>
                    <h2 className='center'>{heading} Coupon Details</h2>
                        <div>
                            <label>Coupon Code: </label>
                            <input
                                type='text'
                                value={couponCode}
                                name="couponCode"
                                onChange={this.onChange}
                            />
                        </div>
                        <div>
                            <label>Discount: </label>
                            <input
                                type='Number'
                                value={discount}
                                name="discount"
                                onChange={this.onChange}
                            />
                        </div>
                        <div>
                            <label>Issue Date: </label>
                            <input
                                type='Date'
                                value={issueDate}
                                name="issueDate"
                                onChange={this.onChange}
                            />
                        </div>
                        <div>
                            <label>Expiry Date: </label>
                            <input
                                type='Date'
                                value={expiryDate}
                                name="expiryDate"
                                onChange={this.onChange}
                            />
                        </div>
                    
                        <center>
                            <button onClick={heading === "Edit " ? this.editCoupon : this.saveCoupon}>{heading}</button>
                        </center>
                </form>
            </div>
        )
        // if(couponCode==='' && discount==='' && issueDate==='' && expiryDate==='')
        //     temp=true;
    }
}
AddCoupon.propTypes={
    createCoupon:PropTypes.func,
    coupons:PropTypes.array,
    updateCoupon:PropTypes.func
}
export default AddCoupon;