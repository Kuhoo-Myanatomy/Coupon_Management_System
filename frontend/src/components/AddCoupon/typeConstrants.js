export const mockItem={
    couponCode:"DEMO10",
    discount: 100,
    
}