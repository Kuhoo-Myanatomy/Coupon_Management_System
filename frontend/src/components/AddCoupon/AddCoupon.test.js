import Enzyme, { shallow } from 'enzyme'
import Adapter from '@cfaester/enzyme-adapter-react-18';
import AddCoupon from './AddCoupon';

Enzyme.configure({ adapter: new Adapter() })

describe("Testing AddCoupon Component",()=>{
    let component;
    beforeEach(()=>{
        component=shallow(<AddCoupon/>);
    })

    it("Matching Addcoupon's snapshot",()=>{
        expect(component).toMatchSnapshot();
    });

    it("check if couponCode input is clicked and state flag updated",()=>{
        expect(component.state("couponCodeClicked")).toBe(false);
        component.find('input[name="couponCode"]').simulate('click');
        expect(component.state("couponCodeClicked")).toBe(true);
        component.setState({ couponCode: 'TEST100'});
        expect(component.state("couponCode")).toBe("TEST100");
    })

    it("check title message if button diabled",()=>{
        expect(component.find('button').prop('disabled')).toBe(true);
        expect(component.find('button').prop('title')).toBe("Please fill the details to enable the submit button");
        component.setState({ couponCode: 'TEST100', discount: 10, issueDate: '2023-04-29', expiryDate: '2023-05-01' });
        expect(component.find('button').prop('disabled')).toBe(false);
        expect(component.find('button').prop('title')).toBe("");
    })


    it("disable submit button if fields are empty else enabled", () => {
        expect(component.find('button').prop('disabled')).toBe(true);
        expect(component.find('button').prop('disabled')).toBe(true);
        component.setState({ couponCode: 'TEST100', discount: 10, issueDate: '2023-04-29' });
        expect(component.find('button').prop('disabled')).toBe(true);
        component.setState({ couponCode: 'TEST100', discount: 10, issueDate: '2023-04-29', expiryDate: '2023-05-21' });
        expect(component.find('button').prop('disabled')).toBe(false);
    });

    it('check error message displayed when couponCode is clicked and left empty', () => {
        expect(component.find('p').length).toBe(0);
        component.find('input[name="couponCode"]').simulate('click');
        expect(component.find('p').length).toBe(1);
        expect(component.find('p').text()).toBe("This field can't be left blank");
        component.find('input[name="couponCode"]').simulate('click');
        component.setState({ couponCode: 'TEST100'});
        expect(component.find('p').length).toBe(0);
    });


    it('error message if discount is clicked and written negative or 0 value', () => {
        component.find('input[name="discount"]').simulate('click');
        
        component.setState({ discount: -10 });
        expect(component.find('p').length).toBe(1);
        expect(component.find('p').text()).toBe("Discount should be greater than 0");
        component.find('input[name="couponCode"]').simulate('click');
        component.setState({ discount: -10 ,couponCode:"  "});
        expect(component.find('p').length).toBe(2);
        // expect(component.find('p').text()).toBe("Discount should be greater than 0");
    });


    it('error message displayed when issueDate is clicked and left empty', () => {
        expect(component.find('p').length).toBe(0);
        component.find('input[name="issueDate"]').simulate('click');
        expect(component.find('p').length).toBe(1);
        expect(component.find('p').text()).toBe("This field can't be left blank");
        component.find('input[name="issueDate"]').simulate('click');
        component.setState({ issueDate: '2023-04-29'});
        expect(component.find('p').length).toBe(0);
    });


    it('check error for expiryDate is clicked and left blank', () => {
        expect(component.find('p').length).toBe(0);
        component.find('input[name="expiryDate"]').simulate('click');
        expect(component.find('p').length).toBe(1);
        expect(component.find('p').text()).toBe("This field can't be left blank");
        component.find('input[name="expiryDate"]').simulate('click');
        component.setState({ expiryDate: '2023-05-29'});
        expect(component.find('p').length).toBe(0);
    });

    it("check if add button is clicked and confirm dialog opens",()=>{
        const fun=jest.fn();
        window.confirm=fun;
        expect(fun).toHaveBeenCalledTimes(0);
        component.find('button[type="submit"]').props().onClick();
        expect(fun).toHaveBeenCalledTimes(1);
    })

    it('saveCoupon function call when Add button and confirm in dialog box', () => {
        const saveCouponMock = jest.fn(); //created a function
        component.instance().saveCoupon = saveCouponMock; //assigned it to saveCoupon() of component instance
        window.confirm = jest.fn(() => true); //setting always true value to window.confirm
        component.find('button[type="submit"]').simulate('click'); 
        expect(saveCouponMock).toHaveBeenCalled(); 
    });


    it('saveCoupon function call when Add button and cancel in confirm dialog box', () => {
        const saveCouponMock = jest.fn();
        component.instance().saveCoupon = saveCouponMock;
        window.confirm = jest.fn(() => false);
        component.find('button[type="submit"]').simulate('click');
        expect(saveCouponMock).not.toHaveBeenCalled();
    });
})