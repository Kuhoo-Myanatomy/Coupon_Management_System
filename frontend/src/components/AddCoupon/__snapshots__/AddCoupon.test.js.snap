// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AddCoupon Component Matching Addcoupon's snapshot 1`] = `ShallowWrapper {}`;
